uncoveredIntervals can overflow query response header

CachingClusteredClient builds up a list of "uncovered intervals", intervals within the query interval that had no data in the underlying segments.

This list is returned as part of the response header in the "X-Druid-Response-Context" field.

If there are a large number of uncovered intervals, the response header may not be large enough to hold this list. On my local system I saw the header buffer had an 8KB size limit.

I noticed this while running a SegmentMetadataQuery on a set of minute-granularity segments that only contained data for even minutes.

e.g.
```
2015-12-16T17:33:21,646 WARN [qtp296974277-44] org.eclipse.jetty.server.HttpChannel - /druid/v2/?pretty
java.io.IOException: Response header too large
	at org.eclipse.jetty.http.HttpGenerator.generateResponse(HttpGenerator.java:400) ~[jetty-http-9.2.5.v20141112.jar:9.2.5.v20141112]
	at org.eclipse.jetty.server.HttpConnection$SendCallback.process(HttpConnection.java:599) ~[jetty-server-9.2.5.v20141112.jar:9.2.5.v20141112]
	at org.eclipse.jetty.util.IteratingCallback.processing(IteratingCallback.java:246) ~[jetty-util-9.2.5.v20141112.jar:9.2.5.v20141112]
	at org.eclipse.jetty.util.IteratingCallback.iterate(IteratingCallback.java:208) ~[jetty-util-9.2.5.v20141112.jar:9.2.5.v20141112]
	at org.eclipse.jetty.server.HttpConnection.send(HttpConnection.java:448) ~[jetty-server-9.2.5.v20141112.jar:9.2.5.v20141112]
	at org.eclipse.jetty.server.HttpChannel.sendResponse(HttpChannel.java:762) ~[jetty-server-9.2.5.v20141112.jar:9.2.5.v20141112]
	at org.eclipse.jetty.server.HttpChannel.write(HttpChannel.java:800) ~[jetty-server-9.2.5.v20141112.jar:9.2.5.v20141112]
	at org.eclipse.jetty.server.HttpOutput.write(HttpOutput.java:139) ~[jetty-server-9.2.5.v20141112.jar:9.2.5.v20141112]
	at org.eclipse.jetty.server.HttpOutput.write(HttpOutput.java:132) ~[jetty-server-9.2.5.v20141112.jar:9.2.5.v20141112]
	at org.eclipse.jetty.server.HttpOutput.write(HttpOutput.java:347) ~[jetty-server-9.2.5.v20141112.jar:9.2.5.v20141112]
...
Caused by: java.nio.BufferOverflowException
	at java.nio.Buffer.nextPutIndex(Buffer.java:521) ~[?:1.8.0_51]
	at java.nio.HeapByteBuffer.put(HeapByteBuffer.java:169) ~[?:1.8.0_51]
	at org.eclipse.jetty.http.HttpGenerator.putSanitisedValue(HttpGenerator.java:1057) ~[jetty-http-9.2.5.v20141112.jar:9.2.5.v20141112]
	at org.eclipse.jetty.http.HttpGenerator.putTo(HttpGenerator.java:1079) ~[jetty-http-9.2.5.v20141112.jar:9.2.5.v20141112]
	at org.eclipse.jetty.http.HttpGenerator.generateHeaders(HttpGenerator.java:703) ~[jetty-http-9.2.5.v20141112.jar:9.2.5.v20141112]
	at org.eclipse.jetty.http.HttpGenerator.generateResponse(HttpGenerator.java:385) ~[jetty-http-9.2.5.v20141112.jar:9.2.5.v20141112]
	... 74 more
```

--- a/processing/src/main/java/io/druid/query/BaseQuery.java
+++ b/processing/src/main/java/io/druid/query/BaseQuery.java
@@ -62,6 +62,11 @@ public abstract class BaseQuery<T extends Comparable<T>> implements Query<T>
     return parseBoolean(query, "finalize", defaultValue);
   }
 
+  public static <T> int getContextUncoveredIntervalsLimit(Query<T> query, int defaultValue)
+  {
+    return parseInt(query, "uncoveredIntervalsLimit", defaultValue);
+  }
+
   private static <T> int parseInt(Query<T> query, String key, int defaultValue)
   {
     Object val = query.getContextValue(key);

--- a/server/src/main/java/io/druid/client/CachingClusteredClient.java
+++ b/server/src/main/java/io/druid/client/CachingClusteredClient.java
@@ -166,33 +166,54 @@ public class CachingClusteredClient<T> implements QueryRunner<T>
     Set<Pair<ServerSelector, SegmentDescriptor>> segments = Sets.newLinkedHashSet();
 
     List<TimelineObjectHolder<String, ServerSelector>> serversLookup = Lists.newLinkedList();
-    List<Interval> uncoveredIntervals = Lists.newLinkedList();
 
-    for (Interval interval : query.getIntervals()) {
-      Iterable<TimelineObjectHolder<String, ServerSelector>> lookup = timeline.lookup(interval);
-      long startMillis = interval.getStartMillis();
-      long endMillis = interval.getEndMillis();
-      for (TimelineObjectHolder<String, ServerSelector> holder : lookup) {
-        Interval holderInterval = holder.getInterval();
-        long intervalStart = holderInterval.getStartMillis();
-        if (startMillis != intervalStart) {
-          uncoveredIntervals.add(new Interval(startMillis, intervalStart));
+    // Note that enabling this leads to putting uncovered intervals information in the response headers
+    // and might blow up in some cases https://github.com/druid-io/druid/issues/2108
+    int uncoveredIntervalsLimit = BaseQuery.getContextUncoveredIntervalsLimit(query, 0);
+
+    if (uncoveredIntervalsLimit > 0) {
+      List<Interval> uncoveredIntervals = Lists.newArrayListWithCapacity(uncoveredIntervalsLimit);
+      boolean uncoveredIntervalsOverflowed = false;
+
+      for (Interval interval : query.getIntervals()) {
+        Iterable<TimelineObjectHolder<String, ServerSelector>> lookup = timeline.lookup(interval);
+        long startMillis = interval.getStartMillis();
+        long endMillis = interval.getEndMillis();
+        for (TimelineObjectHolder<String, ServerSelector> holder : lookup) {
+          Interval holderInterval = holder.getInterval();
+          long intervalStart = holderInterval.getStartMillis();
+          if (!uncoveredIntervalsOverflowed && startMillis != intervalStart) {
+            if (uncoveredIntervalsLimit > uncoveredIntervals.size()) {
+              uncoveredIntervals.add(new Interval(startMillis, intervalStart));
+            } else {
+              uncoveredIntervalsOverflowed = true;
+            }
+          }
+          startMillis = holderInterval.getEndMillis();
+          serversLookup.add(holder);
         }
-        startMillis = holderInterval.getEndMillis();
-        serversLookup.add(holder);
+
+        if (!uncoveredIntervalsOverflowed && startMillis < endMillis) {
+          if (uncoveredIntervalsLimit > uncoveredIntervals.size()) {
+            uncoveredIntervals.add(new Interval(startMillis, endMillis));
+          } else {
+            uncoveredIntervalsOverflowed = true;
+          }
+        }
       }
 
-      if (startMillis < endMillis) {
-        uncoveredIntervals.add(new Interval(startMillis, endMillis));
+      if (!uncoveredIntervals.isEmpty()) {
+        // This returns intervals for which NO segment is present.
+        // Which is not necessarily an indication that the data doesn't exist or is
+        // incomplete. The data could exist and just not be loaded yet.  In either
+        // case, though, this query will not include any data from the identified intervals.
+        responseContext.put("uncoveredIntervals", uncoveredIntervals);
+        responseContext.put("uncoveredIntervalsOverflowed", uncoveredIntervalsOverflowed);
       }
-    }
-
-    if (!uncoveredIntervals.isEmpty()) {
-      // This returns intervals for which NO segment is present.
-      // Which is not necessarily an indication that the data doesn't exist or is
-      // incomplete. The data could exist and just not be loaded yet.  In either
-      // case, though, this query will not include any data from the identified intervals.
-      responseContext.put("uncoveredIntervals", uncoveredIntervals);
+    } else {
+      for (Interval interval : query.getIntervals()) {
+        Iterables.addAll(serversLookup, timeline.lookup(interval));
+      }
     }
 
     // Let tool chest filter out unneeded segments

--- a/server/src/test/java/io/druid/client/CachingClusteredClientFunctionalityTest.java
+++ b/server/src/test/java/io/druid/client/CachingClusteredClientFunctionalityTest.java
@@ -18,6 +18,7 @@
  */
 package io.druid.client;
 
+import com.google.common.collect.ImmutableMap;
 import com.google.common.collect.Lists;
 import com.google.common.collect.Ordering;
 import com.google.common.util.concurrent.ListeningExecutorService;
@@ -76,12 +77,14 @@ public class CachingClusteredClientFunctionalityTest {
   public void testUncoveredInterval() throws Exception {
     addToTimeline(new Interval("2015-01-02/2015-01-03"), "1");
     addToTimeline(new Interval("2015-01-04/2015-01-05"), "1");
+    addToTimeline(new Interval("2015-02-04/2015-02-05"), "1");
 
     final Druids.TimeseriesQueryBuilder builder = Druids.newTimeseriesQueryBuilder()
         .dataSource("test")
         .intervals("2015-01-02/2015-01-03")
         .granularity("day")
-        .aggregators(Arrays.<AggregatorFactory>asList(new CountAggregatorFactory("rows")));
+        .aggregators(Arrays.<AggregatorFactory>asList(new CountAggregatorFactory("rows")))
+        .context(ImmutableMap.<String, Object>of("uncoveredIntervalsLimit", 3));
 
     Map<String, Object> responseContext = new HashMap<>();
     client.run(builder.build(), responseContext);
@@ -90,45 +93,51 @@ public class CachingClusteredClientFunctionalityTest {
     builder.intervals("2015-01-01/2015-01-03");
     responseContext = new HashMap<>();
     client.run(builder.build(), responseContext);
-    assertUncovered(responseContext, "2015-01-01/2015-01-02");
+    assertUncovered(responseContext, false, "2015-01-01/2015-01-02");
 
     builder.intervals("2015-01-01/2015-01-04");
     responseContext = new HashMap<>();
     client.run(builder.build(), responseContext);
-    assertUncovered(responseContext, "2015-01-01/2015-01-02", "2015-01-03/2015-01-04");
+    assertUncovered(responseContext, false, "2015-01-01/2015-01-02", "2015-01-03/2015-01-04");
 
     builder.intervals("2015-01-02/2015-01-04");
     responseContext = new HashMap<>();
     client.run(builder.build(), responseContext);
-    assertUncovered(responseContext, "2015-01-03/2015-01-04");
+    assertUncovered(responseContext, false, "2015-01-03/2015-01-04");
 
     builder.intervals("2015-01-01/2015-01-30");
     responseContext = new HashMap<>();
     client.run(builder.build(), responseContext);
-    assertUncovered(responseContext, "2015-01-01/2015-01-02", "2015-01-03/2015-01-04", "2015-01-05/2015-01-30");
+    assertUncovered(responseContext, false, "2015-01-01/2015-01-02", "2015-01-03/2015-01-04", "2015-01-05/2015-01-30");
 
     builder.intervals("2015-01-02/2015-01-30");
     responseContext = new HashMap<>();
     client.run(builder.build(), responseContext);
-    assertUncovered(responseContext, "2015-01-03/2015-01-04", "2015-01-05/2015-01-30");
+    assertUncovered(responseContext, false, "2015-01-03/2015-01-04", "2015-01-05/2015-01-30");
 
     builder.intervals("2015-01-04/2015-01-30");
     responseContext = new HashMap<>();
     client.run(builder.build(), responseContext);
-    assertUncovered(responseContext, "2015-01-05/2015-01-30");
+    assertUncovered(responseContext, false, "2015-01-05/2015-01-30");
 
     builder.intervals("2015-01-10/2015-01-30");
     responseContext = new HashMap<>();
     client.run(builder.build(), responseContext);
-    assertUncovered(responseContext, "2015-01-10/2015-01-30");
+    assertUncovered(responseContext, false, "2015-01-10/2015-01-30");
+
+    builder.intervals("2015-01-01/2015-02-25");
+    responseContext = new HashMap<>();
+    client.run(builder.build(), responseContext);
+    assertUncovered(responseContext, true, "2015-01-01/2015-01-02", "2015-01-03/2015-01-04", "2015-01-05/2015-02-04");
   }
 
-  private void assertUncovered(Map<String, Object> context, String... intervals) {
+  private void assertUncovered(Map<String, Object> context, boolean uncoveredIntervalsOverflowed, String... intervals) {
     List<Interval> expectedList = Lists.newArrayListWithExpectedSize(intervals.length);
     for (String interval : intervals) {
       expectedList.add(new Interval(interval));
     }
     Assert.assertEquals((Object) expectedList, context.get("uncoveredIntervals"));
+    Assert.assertEquals(uncoveredIntervalsOverflowed, context.get("uncoveredIntervalsOverflowed"));
   }
 
   private void addToTimeline(Interval interval, String version) {

